### PR TITLE
Fix MakeRelative regression in v16.10

### LINKED ISSUE #6493: $([MSBuild]::MakeRelative()) returns different values on 16.10 than older version
### Issue Description
Newest version of VisualStudio (MSBuild 16.10) return different values on msbuild property function MakeRelative() with paths ending backslash.

### Steps to Reproduce
code: $([MSBuild]::MakeRelative('c:\windows\\', 'c:\windows\system32\\'))
MSBuild 16.10 returns:  'system32'
MSBuild 16.9 and older returns: 'system32\\'

### Expected Behavior
according to https://docs.microsoft.com/en-us/dotnet/api/system.uri.makerelativeuri?view=net-5.0
it should returns 'system32\\'

### Actual Behavior
returns 'system32'

### Analysis
MSBuild unittests does not contains paths ending backslash
https://github.com/dotnet/msbuild/blob/main/src/Shared/UnitTests/FileUtilities_Tests.cs#L86

### Versions & Configurations
16.10.0.26302

### PR DESCRIPTION
**Pull request by dsparkplug**
Fixes #6493

Replaces PR by @pmisik : https://github.com/dotnet/msbuild/pull/6504

### Context

Recent change to MakeRelative function was causing existing targets to fail. The change removed the backslash from the end of paths. This change causes the function to revert to the previous behaviour.

### Changes Made

FileUtilities.cs updated

### Testing

Added unit tests to FileUtilities_Tests.cs.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/msbuild?pullRequest=6508) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/msbuild?pullRequest=6508) before we can accept your contribution.<br/><br/>:x: dsparkplug [sign now](https://cla.dotnetfoundation.org/dotnet/msbuild?pullRequest=6508)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/msbuild?pullRequest=6508) it.</sub>

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/msbuild?pullRequest=6508) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/msbuild?pullRequest=6508) before we can accept your contribution.<br/><br/>:x: dsparkplug [sign now](https://cla.dotnetfoundation.org/dotnet/msbuild?pullRequest=6508)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/msbuild?pullRequest=6508) it.</sub>

File: src/Shared/UnitTests/FileUtilities_Tests.cs [hunk 1]
```diff
@@ -97,6 +97,12 @@ public void MakeRelativeTests()
                 Assert.Equal(@"\\host\path\file", FileUtilities.MakeRelative(@"c:\abc\def", @"\\host\path\file"));
                 Assert.Equal(@"\\host\d$\file", FileUtilities.MakeRelative(@"c:\abc\def", @"\\host\d$\file"));
                 Assert.Equal(@"..\fff\ggg.hh", FileUtilities.MakeRelative(@"c:\foo\bar\..\abc\cde", @"c:\foo\bar\..\abc\fff\ggg.hh"));
+
+                /* Directories */
+                Assert.Equal(@"def\", FileUtilities.MakeRelative(@"c:\abc\", @"c:\abc\def\"));
+                Assert.Equal(@"..\", FileUtilities.MakeRelative(@"c:\abc\def\xyz\", @"c:\abc\def\"));
+                Assert.Equal(@"..\ttt\", FileUtilities.MakeRelative(@"c:\abc\def\xyz\", @"c:\abc\def\ttt\"));
+                Assert.Equal(@".", FileUtilities.MakeRelative(@"c:\abc\def\", @"c:\abc\def\"));
```
Comment (pmisik) [inline src/Shared/UnitTests/FileUtilities_Tests.cs:105]: There is opportunity to extend test coverage for combination of file+directory and directory+file.
```
/* Directory + File */
 Assert.Equal(@"def", FileUtilities.MakeRelative(@"c:\abc\", @"c:\abc\def"));
 Assert.Equal(@"..\..\def", FileUtilities.MakeRelative(@c:\abc\def\xyz\", @"c:\abc\def"));
 Assert.Equal(@"..\ghi", FileUtilities.MakeRelative(@c:\abc\def\xyz\", @"c:\abc\def\ghi"));
 Assert.Equal(@"..\def", FileUtilities.MakeRelative(@c:\abc\def\", @"c:\abc\def"));

/* File + Directory */
 Assert.Equal(@"def\", FileUtilities.MakeRelative(@c:\abc", @"c:\abc\def\"));
 Assert.Equal(@"..\", FileUtilities.MakeRelative(@c:\abc\def\xyz", @"c:\abc\def\"));
 Assert.Equal(@"..\ghi\", FileUtilities.MakeRelative(@c:\abc\def\xyz", @"c:\abc\def\ghi\"));
 Assert.Equal(@"", FileUtilities.MakeRelative(@c:\abc\def", @"c:\abc\def\"));
```

Comment (dsparkplug) [reply src/Shared/UnitTests/FileUtilities_Tests.cs:105]: Good suggestion - have added more tests

**Comment by dsparkplug** [comment]
Hmmm.. didn't expect all those commits to be added to the pull request when changing the base branch. Hopefully it's all good still?

File: src/Shared/FileUtilities.cs [hunk 2]
```diff
@@ -1078,7 +1078,13 @@ internal static string MakeRelative(string basePath, string path)
             {
                 sb.Append(splitPath[i]).Append(Path.DirectorySeparatorChar);
             }
-            sb.Length--;
+
+            int length = fullPath.Length;
+            if (length < 1 || fullPath[length - 1] != Path.DirectorySeparatorChar)
```
Comment (Forgind) [inline src/Shared/FileUtilities.cs:1083]:
```suggestion
            if (fullPath[fullPath.Length - 1] != Path.DirectorySeparatorChar)
```

length is a little too generic to be useful, but this is a fairly common idiom, so I'd just simplify it like this.

Also, I don't think we need to check length < 1, do we? Can Path.GetFullPath return something of length 0?

**Review by Forgind — APPROVED** [review]
Looks good! I think I know when this came in—the original uri-based implementation failed tests in the installer repo, for some reason. This version passed those tests and was a bit faster—but also inadvertently changed our behavior with slashes. I don't see why that would be necessary to maintain, so I think this is a good change.

**Comment by Forgind** [comment]
Oh, and also:
https://github.com/dotnet/msbuild/pull/6504#issuecomment-852406031

🙂

Is there any reason to keep #6504 around anymore?

**Comment by dsparkplug** [comment]
The other pr was started by @pmisik - it looks like he has pulled my changes into that and rebased to v16.10 - but it doesn't build for some reason. I really should have based this pr on v16.10 as the fix is needed to fix major dependency issues. Will attempt to base it on v16.10 do this now - do I need to rebase the source before using the new update to base branch feature?

**Comment by dsparkplug** [comment]
Have created a new PR https://github.com/dotnet/msbuild/pull/6513 with the same changes squashed and rebased onto vs16.10 branch

**Pull request closed without merging**